File: app/src/main/java/com/affiliatehunter/utils/Scoring.kt
package com.affiliatehunter.utils
import kotlin.math.log10
object Scoring {
    fun score(sold:Int, rating:Double, price:Long, velocityFactor:Double=1.0, trustFactor:Double=1.0): Double {
        val hargaFactor = if(price in 50000..300000) 1.0 else 0.7
        return log10((sold+1).toDouble()) * rating * hargaFactor * velocityFactor * trustFactor
    }
    fun commission(price:Long, rate:Double): Long = (price * rate).toLong()
    fun velocityFlag(soldPerDay: Double, oldPerDay: Double?): String? {
        if(oldPerDay==null || oldPerDay<=0) return null
        val diff = (soldPerDay - oldPerDay)/oldPerDay
        return if(diff >= 0.20) "+${(diff*100).toInt()}% 7d" else null
    }
    fun trustFlag(shopAgeDays:Int, reviewCount:Int, rating:Double): Pair<Boolean,String?> {
        if(shopAgeDays in 1..29) return true to "toko ${shopAgeDays} hari"
        if(rating>=4.9 && reviewCount <10) return true to "rating 4.9 tapi ulasan <10"
        return false to null
    }
    fun rateForCategory(cat:String): Double = when(cat.lowercase()){
        "parfum"->0.08; "skincare"->0.10; "hijab"->0.05; "jam"->0.07; "tas"->0.08; "sepatu"->0.07; else->0.08
    }
}

File: web/lib/scoring.ts
export function score(sold:number, rating:number, price:number){
  const hargaFactor = price>=50000 && price<=300000 ? 1 : 0.7
  return Math.log10(sold+1) * rating * hargaFactor
}
export function commission(price:number, rate:number){ return Math.round(price*rate) }
export function rateForCategory(cat:string){
  const c=cat.toLowerCase()
  if(c.includes("parfum")) return 0.08
  if(c.includes("skincare")||c.includes("skin")) return 0.10
  if(c.includes("hijab")) return 0.05
  if(c.includes("jam")) return 0.07
  if(c.includes("tas")) return 0.08
  if(c.includes("sepatu")) return 0.07
  return 0.08
}
export function trustFlag(shopAgeDays:number, reviewCount:number, rating:number): [boolean, string|null]{
  if(shopAgeDays>=1 && shopAgeDays<=29) return [true, `toko ${shopAgeDays} hari`]
  if(rating>=4.9 && reviewCount<10) return [true, "rating 4.9 tapi ulasan <10"]
  return [false, null]
}
export type Product = {
  id:string; name:string; price:number; sold:number; rating:number; reviewCount:number;
  shopName:string; shopAgeDays:number; imageUrl:string|null; link:string; category:string;
  soldPerDay:number; commissionRate:number; commissionEst:number; score:number; isOP:boolean;
  isFlagged:boolean; flagReason:string|null; velocityBadge?:string|null;
}
